### ISSUE TITLE
Move/edit whiteboard objects

### ISSUE THREAD

**Issue by bigbluebutton-issue-import** [bigbluebutton/bigbluebutton#2351]
Originally reported on Google Code with ID 1599

```
Currently, when the presenter draws on the whiteboard, the object is fixed.  For example,
they can't move/resize the square after it is drawn, nor can they edit a text field.

Add editing of objects to the whiteboard.
```

Reported by `ffdixon` on 2013-07-28 14:24:37

**Comment by poetslipvis** [comment]
Please don't forget about this, it's 4 years old and hasn't been implemented, maybe up the priority. It would be great to be able to move elements on the whiteboard and to be able to edit text. Every conference I give I need to rewrite text...

**Comment by nordby** [comment]
maybe now something has changed?

**Comment by ffdixon** [comment]
We saw your similar comment in #7036.  We'll update this issue when we have more details on the what is planned for the next release of BigBlueButton, which will be version 2.3.

**Comment by fiq** [comment]
+1 for this feature. I've just had some feedback that not being able to move an object is seen as a limitation when delivering collaborative exercises such as sorting/grouping. It would _also_ be useful for the whiteboard to be fist class as opposed to an object which depends on being overlayed over a white slide. We're measuring it against google-jamboard which is quite similar except that it's slide-independent and allows you to move content.

**Comment by lukas71** [comment]
+1 for that. And more the possibility of clip-art and pre-configured image to import...

**Comment by iSamof** [comment]
Friends - the enhancements you mention are great BUT google-jamboard cannot do half the stuff that BBB is doing. And BBB cannot be all things to everyone, all at the same time. 

Drawing and manipulating objects can be done in a variety of tools including Power Point on a shared screen - the presenter would have to move the objects around. 

From my business viewpoint, BBB is a first class teaching and conferencing tool and needs to build on that. There is so much to be done that focus has to be maintained. The white boarding works for us ... and if BBB upgrade it to vector objects then great. 

But I'd prefer energy and focus to be devoted to what makes BBB unique and not diluted to include tools that others already provide. I get it, those others are not integrated. 

Just my 2 cents worth ;-)

**Comment by poetslipvis** [comment]
We have over 100 instructors using BBB and the very limited whiteboard functions have not gone unnoticed. The fact you can't drop and drag or edit is quite basic.

**Comment by ProfMarte** [comment]
Hi,
do you know https://github.com/OpenBoard-org/OpenBoard ?
May be you can share code from there to integrate in BBB. OpenBoard is one of the greatest Whiteboard tools and it is open source. May be it is possible to cooperate. 
That's just I think: Don't try to invent the wheel new.
My workaround is to share the screen by working in OpenBoard, but then it is no colaboration possible. On the other hand: WIthout the possibility to delete selectable objects (at least having a rubber tool like in paint) and without the possibility to move objets, colaborative use is very very limited and unthinkable while teaching. 
May be this idea helps. 
Regards
Marte

**Comment by Beiri22** [comment]
We also suffer from the limited capabilities of the whiteboard. Is there any way to reactivate this issue; and increase its priority?

**Comment by ferdinandSoethe** [comment]
+1 for that

**Comment by matiasilva** [comment]
Hi guys. If you are or know any developers, then pull requests are really helpful. Time is limited and it takes time to test and develop features. The community can't just switch priorities when certain users demand a feature. If you really want to see a feature, you can either reach out to the companies providing commercial support for BigBlueButton or start a discussion in the developers group about how to go implementing it. Certainly more useful than stating what's wrong with the current implementation and expecting others to magically develop the feature.

As for my opinion on this, I agree with @iSamof. Working on embedding such a whiteboard tool would be ideal. For the time being, I find using Google JAMboard and sharing my screen for cases where I need extra functionality works just fine.

**Comment by ferdinandSoethe** [comment]
As a former committer to Apache projects I can see your point, @matiasilva. However if you are a teacher at a German University you are not allowed to use Google or other external solutions because of Data Protections laws. And using screen sharing with a larger number of users quickly becomes a performance issue als well. So improving the well designed presentation system makes sense and might be yet another step towards making BBB the best system there is.

**Comment by sonerd** [comment]
+1 for this feature

**Comment by mehdikazemi3120** [comment]
+1 for this

**Comment by ffdixon** [comment]
We appreciate the feedback. We've done some performance improvements in the whiteboard for 2.3, but have no plans to enable editing/moving of objects in that release. We're really trying to keep the scope down on 2.3 so we can ship a stable build as soon as possible (on Ubuntu 18.04) and build more features upon it.

**Comment by Rortar** [comment]
Understood that it won't be in 2.3, but would be a nice one for the roadmap +1

**Comment by MrCapitalist** [comment]
+1 for this

I saw this amazing feature in another web conference application (I think it was OpenMeeting).
The objects was draggable, editable and user could resize them or change properties of them. For example for texts could change text color, size or font.

In that platform everything was Web-based and same as BigBlueButton I just need an address of website without install any prerequired software

**Comment by amishbhavsar** [comment]
Would love to see better whiteboard annotation features in BigBlueButton. I have been waiting for this enhancement to
1. edit drawn objects, 
2. erase them and 
3. also move / rotate them 
This will greatly enhance the use of BBB for education purposes.
I will eagerly wait for this enhancement.

**Comment by ffdixon** [comment]
We're keen to implement this as well, but it's going be after the 2.3 release which will upgrade a lot of the key components and give us a foundation on which to build this feature.

**Comment by amishbhavsar** [comment]
Now that 2.3 is released. Any idea on when this feature will be made available on BBB?

**Comment by MrCapitalist** [comment]
Hi BigBlueButton powerful team!
+1 for this feature.

And I have some additional requests too:

1. Please add arrow tools to the whiteboard like this: https://s4.uupload.ir/files/111_mjpa.png
2. Please show some way to disable the feature that doesn't allow user text input shows to viewers until the moderator click to an empty area and finish his editing (we don't have this problem in v2.2)
3. Please add the "Bring forward/backward" feature for shapes and other objects on the whiteboard and the ability to change the font of text like this: https://s4.uupload.ir/files/22222_21du.png

Thanks

**Comment by hex-m** [comment]
Thank you for your input @MrCapitalist. We try to keep issues focused by having separate issues for different requests.

Your second suggestion is already covered in https://github.com/bigbluebutton/bigbluebutton/issues/12345.

Please create new issues for the other two.

**Comment by MrCapitalist** [comment]
> Thank you for your input @MrCapitalist. We try to keep issues focused by having separate issues for different requests.
> 
> Your second suggestion is already covered in #12345.
> 
> Please create new issues for the other two.

Thanks. I've just opened feature request issues:
Arrow tools: #13284 
Move object and Bring forward/backward: #13285 
Fonts: #13286

**Comment by solentenglish** [comment]
+1 for this feature

**Comment by hiroshisuga** [comment]
I have missed the post by @MW360 , but I have almost done it independently... Should I continue working on it?

![Animation](https://user-images.githubusercontent.com/45039819/141235688-dcdf5efb-6a61-4ff4-b5a4-6e0db360ddad.gif)

**Comment by MW360** [comment]
Hi @hiroshisuga,
do you plan to implement anything other than the delete function on the whiteboard? 
Would be good to know for my further planning.

So I would implement movement with [react movable components](https://www.npmjs.com/package/react-moveable), scaling of the objects, rotation, moving from background to foreground, etc.
I also plan to make the text boxes editable.

Best regards!

**Comment by hiroshisuga** [comment]
Hi @MW360 , I am now trying to drag and move the drawings, and probably later, scale them. Your solution looks fantastic! I would probably implement the moving function just for fun and post a PR for those who prefer a simple implementation... So please go ahead and propose a PR!

One question: would you expect that the recording function be easily compatible with that React Movable Component?

**Comment by MW360** [comment]
Hi @hiroshisuga,
I am not completely sure about that, as I am in an early stage of development. 
But I expect that there are changes to the recording functionality required.

**Comment by MW360** [comment]
Hi everyone, 
I want to you give a small preliminary roadmap that shows my time table for these features.
Until the end of November I plan to implement the delete functionality of annotations with react moveable. 
After that I want to add new features like moving, scaling, etc. every one or two weeks depending on how challenging the task is.

**Comment by iSamof** [comment]
Thank you Maximilian. This is great work, in addition to Hiroshi's great work.

**Comment by hiroshisuga** [comment]
Implementing the dragging function has almost been finished. Still needs some work but I would say it's promising!
![Uploading Animation.gif…]()

**Comment by hiroshisuga** [comment]
Implementing the dragging function has almost been finished. Still needs some work but I would say it's promising!

![Uploading Animation.gif…]()

**Comment by hiroshisuga** [comment]
Implementing the dragging function has almost been finished. Still needs some work but I would say it's promising!

![Uploading Animation.gif…]()

**Comment by hiroshisuga** [comment]
![Uploading Animation.gif…]()

**Comment by hiroshisuga** [comment]
So far so good. Still needs some work.

![Animation](https://user-images.githubusercontent.com/45039819/142139123-7936831b-d33a-4328-910f-3d2e3a542328.gif)

**Comment by Gigate** [comment]
@MW360 Hi I'm currently working on an advanced eraser implementation with server side clipping based on the work of @hiroshisuga (#13634). I was also planning on adding object manipulation and so on later. It would be great to talk to you to. I'm a computer science student at Ulm University Germany. If you like hit me up via Mail tim-3.schneider@uni-ulm.de

**Comment by hiroshisuga** [comment]
Hi @Gigate , Hi @MW360, great to have multiple projects for WB enhancement (after a long time of no progress)! 
My interest is to make BBB more comfortable for teachers and students, not to re-invent Adobe Illustrator;-.  I am pretty sure you guys are the same. 

So I wonder how we proceed. Ideally, we may collaborate to publish one PR. But my project seems to have already proceeded so much that merging them might be difficult. In addition, the time difference between Germany and Japan will not be ignorable. Thus I decide to post my PR, accepting feedbacks from users including you. Note that this does not mean that I am against your projects. In fact I am very keen to see your solution!

For designing policy, I would
1. keep the eraser and marker function that I posted already, as using a rectangle with background picture would still be useful to precisely modify the drawing. 
2. implement the selecting annotation function, learning from the PR by @Gigate , to move, remove, and bring back and forward the drawings; combining with the function in 1, the teachers would basically be able to do whatever they want.
3. not implement scaling and rotating drawings, as they are not so useful for the learning activities (in my opinion); if you want to scale a drawing, you can delete it and redraw a new one.
4. enable the recording function from the beginning, otherwise I may need to massively change the code later (the current recording function is working in a bit different way as the html5 client and the backward compatibility is also an issue)

For the 3, implementing rotation and scaling is not so difficult as long as we use keyboard shortcuts. But if we want e.g. a corner point to snap the bounding box, it would be very difficult to implement and they will be error-prone. I also think such fancy user interface always make some issue coming from a personal taste (I experienced it when I proposed an animated effect for choosing random user;-). But as always I want to hear from the users and core developers.

**Comment by christianzz** [comment]
@hiroshisuga is there any way we can support (e.g. financially) to make this happen? This is something we are urgently looking for!

**Comment by hiroshisuga** [comment]
Hi @christianzz , I have already submitted the PR (#13769). Feel free to merge it to your local server.

**Comment by antobinary** [comment]
Hi @christianzz 
A couple of the core BBB developers have met with the authors of https://github.com/bigbluebutton/bigbluebutton/pull/13634 (which is building upon @hiroshisuga 's great work) - @Gigate etc - and we hope to see the result of this work land in as part of BBB 2.5. 

We can say exactly the same about @MW360 's upcoming work.

Although there is plenty left to implement, we endorse the approach taken in both areas.

**Comment by czange** [comment]
Hi Anton, thanks for letting me know - sounds promising. We experience, that the BBB whiteboard is neglected and declined by our customers, because it is simply not able to compete with what is out there (MIRO, MURAL etc.) I understand that competing with such specialised offers is difficult - but making things movable might go a long way. We will try to get in touch with those developers ...

**Comment by pascoual** [comment]
As I understand, there is no way to import picture in whiteboard. The goal is to make a simple MURAL, MIRO alternative all in one in BBB for climate sensitization with https://climatefresk.org & https://digitalcollage.org. It need picture (ko) with hidden/shown associated text (ko) and arrow (ok).
If someone as a WIP, please tell me :).

**Comment by hiroshisuga** [comment]
@amishbhavsar , I don't think so, although my latest PR #14540 is compatible to v2.4.x . You can locally deploy it on your server.

**Comment by amishbhavsar** [comment]
Thanks @hiroshisuga .
Can you guide me on how to install this on my instance of BBB?

**Comment by hiroshisuga** [comment]
Hi @amishbhavsar , it's nicely written here: https://docs.bigbluebutton.org/2.4/dev.html
You make your own branch from the PR (or simply download mine, which is https://github.com/hiroshisuga/bigbluebutton/tree/dev2.4.4_selectableAnnotations) , then following the steps above, you deploy both HTML5 client and Akka-Apps.

**Comment by ramonlsouza** [comment]
closed by #14994 (moving/editing object possible in BBB 2.6)
